### PR TITLE
make integration with TMDB api

### PR DESCRIPTION
**Pull request by marekk1992**

### CODE REVIEW

File: src/main/java/com/example/moviemanager/constantValues/MovieType.java [hunk 1]
```diff
@@ -0,0 +1,6 @@
+package com.example.moviemanager.constantValues;
```
Comment (RobertasVainickas) [inline src/main/java/com/example/moviemanager/constantValues/MovieType.java]: It is part of model. Now you have controller layer model classes and service layer model classes but this one is part of both. I suggest having the same with enums

File: src/main/java/com/example/moviemanager/constantValues/MovieType.java [hunk 2]
```diff
@@ -0,0 +1,6 @@
...
+
+public enum MovieType {
+
```
Comment (RobertasVainickas) [inline src/main/java/com/example/moviemanager/constantValues/MovieType.java]: Don't add such blank lines

File: src/main/java/com/example/moviemanager/controller/model/CreateMovieRequest.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+import jakarta.validation.constraints.Min;
+import jakarta.validation.constraints.NotBlank;
+
+public record CreateMovieRequest(
+        @NotBlank(message = "Title is mandatory.")
+        String title,
+
+        @NotBlank(message = "Movie type is mandatory.")
+        @OneOfMovieTypes
+        String type,
```
Comment (RobertasVainickas) [inline src/main/java/com/example/moviemanager/controller/model/CreateMovieRequest.java]: You have separate data type for it. If you use it, you won't even need that custom annotation which validates value correctness

File: src/main/java/com/example/moviemanager/service/model/api/ApiResponse.java [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package com.example.moviemanager.service.model.api;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+
+import java.util.List;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+public record ApiResponse(List<MovieDetailsFromApi> results) {
```
Comment (RobertasVainickas) [inline src/main/java/com/example/moviemanager/service/model/api/ApiResponse.java]: Class name sounds very abstract

File: src/main/java/com/example/moviemanager/service/model/api/MovieDetailsFromApi.java [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package com.example.moviemanager.service.model.api;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+public record MovieDetailsFromApi(String overview, double vote_average) {
```
Comment (RobertasVainickas) [inline src/main/java/com/example/moviemanager/service/model/api/MovieDetailsFromApi.java]: You shouldn't emphasize that details are from api. Simply MovieDetails would be enough in my opinion

File: src/main/java/com/example/moviemanager/service/MovieService.java [hunk 6]
```diff
@@ -1,31 +1,59 @@
...
 import java.util.List;
 import java.util.Optional;
 import java.util.UUID;
+import java.util.stream.Collectors;
 
 @Service
 public class MovieService {
 
+    private static final String TV_SHOWS_URI =
+            "https://api.themoviedb.org/3/search/tv?api_key=77938ea875914a4c8ded6ccb9d840d27&query=";
```
Comment (RobertasVainickas) [inline src/main/java/com/example/moviemanager/service/MovieService.java]: 1. api key should not be hardcoded. It is sensitive information so, you should not expose it anywhere. And it should be configurable through application properties, but should not be added there. 
2. The base URI (https://api.themoviedb.org/) can be configured directly on RestTemplate if you would not create each instance here with RestTemplate() but instead would create a Bean which can be injected. The bean itself can be created with RestTemplateBuilder class. What is more base url should also be configurable through application properties.

File: src/main/java/com/example/moviemanager/service/MovieService.java [hunk 7]
```diff
@@ -36,17 +64,62 @@ public void deleteById(UUID id) {
...
+    private Movie buildMovie(CreateMovieInfo saveMovieRequest, MovieDetailsFromApi movieDetailsFromApi) {
+        return new Movie(
+                saveMovieRequest.title().toUpperCase(),
+                movieDetailsFromApi.overview(),
+                saveMovieRequest.releaseYear(),
+                movieDetailsFromApi.vote_average()
+        );
+    }
+
+    private String buildUri(CreateMovieInfo saveMovieRequest) {
```
Comment (RobertasVainickas) [inline src/main/java/com/example/moviemanager/service/MovieService.java]: URI should be built with UriComponentsBuilder class. It will look much cleaner

File: src/main/java/com/example/moviemanager/service/MovieService.java [hunk 8]
```diff
@@ -36,17 +64,62 @@ public void deleteById(UUID id) {
...
+        String movieType = saveMovieRequest.type();
+        if (movieType.equalsIgnoreCase(MovieType.MOVIE.toString())) {
+            return MOVIES_URI + saveMovieRequest.title() + MOVIE_RELEASE_YEAR + saveMovieRequest.releaseYear();
+        }
+        return TV_SHOWS_URI + saveMovieRequest.title() + TV_SHOW_RELEASE_YEAR + saveMovieRequest.releaseYear();
+    }
+
+    private ApiResponse getResponseFromMovieDatabase(String uri) {
+        try {
+            ApiResponse apiResponse = objectMapper.readValue(
```
Comment (RobertasVainickas) [inline src/main/java/com/example/moviemanager/service/MovieService.java]: No need to use objectMapper, rest template under the hood uses object mapper itself, so just call it with response dto: getForObject(uri, ApiResponse.class)

File: src/main/java/com/example/moviemanager/service/MovieService.java [hunk 9]
```diff
@@ -36,17 +64,62 @@ public void deleteById(UUID id) {
...
+                    new RestTemplate().getForObject(uri, String.class),
+                    ApiResponse.class
+            );
+            if(apiResponse.results().size() != 1) {
+                throw new UniqueMovieNotFoundException("Can`t find unique movie according to your details.");
+            }
+            return apiResponse;
+        } catch (JsonProcessingException e) {
+            throw new FailedToRetrieveDataFromDatabaseException("Failed to retrieve data from database.");
```
Comment (RobertasVainickas) [inline src/main/java/com/example/moviemanager/service/MovieService.java]: JsonProcessingException means that you wasn't able to map response to DTO I think, so your custom exception is a bit misleading

**Review by RobertasVainickas — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/moviemanager/constantValues/MovieType.java (hunk 2 above)
Comment (marekk1992) [reply src/main/java/com/example/moviemanager/constantValues/MovieType.java]: Fixed.

File: src/main/java/com/example/moviemanager/controller/model/CreateMovieRequest.java (hunk 3 above)
Comment (marekk1992) [reply src/main/java/com/example/moviemanager/controller/model/CreateMovieRequest.java]: Fixed.

File: src/main/java/com/example/moviemanager/constantValues/MovieType.java (hunk 1 above)
Comment (marekk1992) [reply src/main/java/com/example/moviemanager/constantValues/MovieType.java]: Moved to service.model package.

File: src/main/java/com/example/moviemanager/service/MovieService.java (hunk 8 above)
Comment (marekk1992) [reply src/main/java/com/example/moviemanager/service/MovieService.java]: Fixed.

File: src/main/java/com/example/moviemanager/service/MovieService.java (hunk 9 above)
Comment (marekk1992) [reply src/main/java/com/example/moviemanager/service/MovieService.java]: Since I removed objectMapper`s readValue() method from code, JsonProcessingException is never thrown. So there is no need to have custom exception. Removed.

File: src/main/java/com/example/moviemanager/service/model/api/ApiResponse.java (hunk 4 above)
Comment (marekk1992) [reply src/main/java/com/example/moviemanager/service/model/api/ApiResponse.java]: Renamed into 'MovieDetailsResponse'.

File: src/main/java/com/example/moviemanager/service/model/api/MovieDetailsFromApi.java (hunk 5 above)
Comment (marekk1992) [reply src/main/java/com/example/moviemanager/service/model/api/MovieDetailsFromApi.java]: Renamed.

File: src/main/java/com/example/moviemanager/service/MovieService.java (hunk 7 above)
Comment (marekk1992) [reply src/main/java/com/example/moviemanager/service/MovieService.java]: Fixed.

File: src/main/java/com/example/moviemanager/service/MovieService.java (hunk 6 above)
Comment (marekk1992) [reply src/main/java/com/example/moviemanager/service/MovieService.java]: 1. Removed API KEY from application properties and updated readme, so that in order to use app, you have to specify api key.
2. Created bean.

File: src/main/java/com/example/moviemanager/service/movieservice/exception/MovieNotFoundException.java [hunk 10]
```diff
@@ -1,4 +1,4 @@
-package com.example.moviemanager.service.exception;
+package com.example.moviemanager.service.movieservice.exception;
```
Comment (RobertasVainickas) [inline src/main/java/com/example/moviemanager/service/movieservice/exception/MovieNotFoundException.java]: I think it was not needed

File: src/main/java/com/example/moviemanager/service/tmdbmovieservice/configuration/ApplicationConfig.java [hunk 11]
```diff
@@ -0,0 +1,15 @@
+package com.example.moviemanager.service.tmdbmovieservice.configuration;
```
Comment (RobertasVainickas) [inline src/main/java/com/example/moviemanager/service/tmdbmovieservice/configuration/ApplicationConfig.java]: instead should be in com.example.moviemanager.configuration package since it is whole application config

File: src/main/java/com/example/moviemanager/service/tmdbmovieservice/configuration/ConfigProperties.java [hunk 12]
```diff
@@ -0,0 +1,36 @@
+package com.example.moviemanager.service.tmdbmovieservice.configuration;
```
Comment (RobertasVainickas) [inline src/main/java/com/example/moviemanager/service/tmdbmovieservice/configuration/ConfigProperties.java]: ditto

File: src/main/java/com/example/moviemanager/service/tmdbmovieservice/configuration/ConfigProperties.java [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+
+import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+@ConfigurationProperties(prefix = "tmdb")
+public class ConfigProperties {
```
Comment (RobertasVainickas) [inline src/main/java/com/example/moviemanager/service/tmdbmovieservice/configuration/ConfigProperties.java]: Even prefix says that it is tmdb properties, so name it accordingly. Because ConfigProperties is very abstract

File: src/main/java/com/example/moviemanager/service/tmdbmovieservice/model/MovieDetails.java [hunk 14]
```diff
@@ -0,0 +1,7 @@
+package com.example.moviemanager.service.tmdbmovieservice.model;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+public record MovieDetails(String overview, double vote_average) {
```
Comment (RobertasVainickas) [inline src/main/java/com/example/moviemanager/service/tmdbmovieservice/model/MovieDetails.java]: vote_average is not named according to Java naming conventions. If you need to provide different naming for json property than it's variable name you can specify it with annotation @JsonProperty("vote_average")

File: src/main/java/com/example/moviemanager/service/tmdbmovieservice/model/MovieType.java [hunk 15]
```diff
@@ -0,0 +1,5 @@
+package com.example.moviemanager.service.tmdbmovieservice.model;
+
+public enum MovieType {
+    MOVIE, TVSHOW;
```
Comment (RobertasVainickas) [inline src/main/java/com/example/moviemanager/service/tmdbmovieservice/model/MovieType.java]: Naming should be according the same conventions like constants, so TV_SHOW

File: src/main/java/com/example/moviemanager/service/tmdbmovieservice/TmdbMovieService.java [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+
+    public Movie getMovie(FindMovieInfo findMovieInfo) {
+        String endpoint = buildEndpoint(findMovieInfo);
+        MovieDetailsResponse movieDetailsResponse = getResponse(endpoint);
+        MovieDetails movieDetails = retrieveMovieDetails(movieDetailsResponse);
+
+        return buildMovie(findMovieInfo, movieDetails);
+    }
+
+    private String buildEndpoint(FindMovieInfo findMovieInfo) {
```
Comment (RobertasVainickas) [inline src/main/java/com/example/moviemanager/service/tmdbmovieservice/TmdbMovieService.java]: buildUri?

File: src/main/java/com/example/moviemanager/service/tmdbmovieservice/TmdbMovieService.java [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+        this.configProperties = configProperties;
+        this.restTemplate = restTemplate;
+    }
+
+    public Movie getMovie(FindMovieInfo findMovieInfo) {
+        String endpoint = buildEndpoint(findMovieInfo);
+        MovieDetailsResponse movieDetailsResponse = getResponse(endpoint);
+        MovieDetails movieDetails = retrieveMovieDetails(movieDetailsResponse);
+
+        return buildMovie(findMovieInfo, movieDetails);
```
Comment (RobertasVainickas) [inline src/main/java/com/example/moviemanager/service/tmdbmovieservice/TmdbMovieService.java]: I think this class has too much responsibility. It should be responsible for integrating with TMDB api and returning the response according the query params, but not building some custom response which is logic behind integration. It is your application's specific logic.

File: src/main/java/com/example/moviemanager/service/tmdbmovieservice/TmdbMovieService.java [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+                .path(PATH + pathComponent)
+                .queryParam(API_KEY, configProperties.getApiKey())
+                .queryParam(QUERY, findMovieInfo.title())
+                .queryParam(releaseYear, findMovieInfo.releaseYear());
+
+        return uri.build().toUriString();
+    }
+
+    private MovieDetailsResponse getResponse(String endpoint) {
+        Optional<MovieDetailsResponse> movieDetailsResponse =
```
Comment (RobertasVainickas) [inline src/main/java/com/example/moviemanager/service/tmdbmovieservice/TmdbMovieService.java]: No purpose of wrapping it with Optional. If it is null then you should check for it later and throw exception

File: src/main/java/com/example/moviemanager/service/tmdbmovieservice/TmdbMovieService.java [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+        String pathComponent = movieType.equals(MovieType.MOVIE) ? MOVIE_TYPE : TV_SHOW_TYPE;
+        String releaseYear = movieType.equals(MovieType.MOVIE) ? MOVIE_RELEASE_YEAR : TV_SHOW_RELEASE_YEAR;
+        UriComponentsBuilder uri = UriComponentsBuilder
+                .newInstance()
+                .path(PATH + pathComponent)
+                .queryParam(API_KEY, configProperties.getApiKey())
+                .queryParam(QUERY, findMovieInfo.title())
+                .queryParam(releaseYear, findMovieInfo.releaseYear());
+
+        return uri.build().toUriString();
```
Comment (RobertasVainickas) [inline src/main/java/com/example/moviemanager/service/tmdbmovieservice/TmdbMovieService.java]: I would simply return the whole chained statement without storing it in variable and later on continue chain operations

File: src/main/java/com/example/moviemanager/service/MovieService.java [hunk 20]
```diff
@@ -49,4 +54,13 @@ public Movie update(UUID id, Movie movie) {
 
         return movieRepository.save(movie);
     }
-}
\ No newline at end of file
+
+    private Movie buildMovie(FindMovieInfo findMovieInfo, MovieDetails movieDetails) {
```
Comment (RobertasVainickas) [inline src/main/java/com/example/moviemanager/service/MovieService.java]: It is not used anywhere

**Review by RobertasVainickas — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/moviemanager/service/MovieService.java (hunk 20 above)
Comment (marekk1992) [reply src/main/java/com/example/moviemanager/service/MovieService.java]: Removed.

File: src/main/java/com/example/moviemanager/service/tmdbmovieservice/TmdbMovieService.java (hunk 16 above)
Comment (marekk1992) [reply src/main/java/com/example/moviemanager/service/tmdbmovieservice/TmdbMovieService.java]: It was at first, but I have changed it later :).

Returned previous name.

File: src/main/java/com/example/moviemanager/service/movieservice/exception/MovieNotFoundException.java (hunk 10 above)
Comment (marekk1992) [reply src/main/java/com/example/moviemanager/service/movieservice/exception/MovieNotFoundException.java]: Removed 'movieservice' package.

File: src/main/java/com/example/moviemanager/service/tmdbmovieservice/TmdbMovieService.java (hunk 19 above)
Comment (marekk1992) [reply src/main/java/com/example/moviemanager/service/tmdbmovieservice/TmdbMovieService.java]: Fixed.

File: src/main/java/com/example/moviemanager/service/tmdbmovieservice/configuration/ApplicationConfig.java (hunk 11 above)
Comment (marekk1992) [reply src/main/java/com/example/moviemanager/service/tmdbmovieservice/configuration/ApplicationConfig.java]: Refactored.

File: src/main/java/com/example/moviemanager/service/tmdbmovieservice/configuration/ConfigProperties.java (hunk 12 above)
Comment (marekk1992) [reply src/main/java/com/example/moviemanager/service/tmdbmovieservice/configuration/ConfigProperties.java]: Refactored.

File: src/main/java/com/example/moviemanager/service/tmdbmovieservice/configuration/ConfigProperties.java (hunk 13 above)
Comment (marekk1992) [reply src/main/java/com/example/moviemanager/service/tmdbmovieservice/configuration/ConfigProperties.java]: Renamed to 'TmdbConfigProperties'.

File: src/main/java/com/example/moviemanager/service/tmdbmovieservice/model/MovieDetails.java (hunk 14 above)
Comment (marekk1992) [reply src/main/java/com/example/moviemanager/service/tmdbmovieservice/model/MovieDetails.java]: Fixed.

File: src/main/java/com/example/moviemanager/service/tmdbmovieservice/TmdbMovieService.java (hunk 18 above)
Comment (marekk1992) [reply src/main/java/com/example/moviemanager/service/tmdbmovieservice/TmdbMovieService.java]: Fixed.

File: src/main/java/com/example/moviemanager/service/tmdbmovieservice/model/MovieType.java (hunk 15 above)
Comment (marekk1992) [reply src/main/java/com/example/moviemanager/service/tmdbmovieservice/model/MovieType.java]: Renamed.

File: src/main/java/com/example/moviemanager/service/tmdbmovieservice/TmdbMovieService.java (hunk 17 above)
Comment (marekk1992) [reply src/main/java/com/example/moviemanager/service/tmdbmovieservice/TmdbMovieService.java]: Refactored. 

TmdbMovieService now returns 'MovieDetails'.

File: src/main/java/com/example/moviemanager/configuration/TmdbConfigProperties.java [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+    public TmdbConfigProperties(String baseUrl, String apiKey) {
+        this.baseUrl = baseUrl;
+        this.apiKey = apiKey;
+    }
+
+    public String getBaseUrl() {
+        return baseUrl;
+    }
+
+    public void setBaseUrl(String baseUrl) {
```
Comment (RobertasVainickas) [inline src/main/java/com/example/moviemanager/configuration/TmdbConfigProperties.java]: Are setters really needed? Or maybe it can be record instead?

File: src/main/java/com/example/moviemanager/service/TmdbMovieService.java [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    private MovieDetails retrieveMovieDetails(MovieDetailsResponse movieDetailsResponse) {
+        if (movieDetailsResponse.results() == null || movieDetailsResponse.results().size() != 1) {
+            throw new UniqueMovieDetailsNotFoundException("Can`t find unique movie according to your request.");
+        }
+        return movieDetailsResponse.results().get(0);
+    }
+
+    private Movie buildMovie(FindMovieInfo findMovieInfo, MovieDetails movieDetails) {
```
Comment (RobertasVainickas) [inline src/main/java/com/example/moviemanager/service/TmdbMovieService.java]: not used anywhere

**Review by RobertasVainickas — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/moviemanager/configuration/TmdbConfigProperties.java (hunk 21 above)
Comment (marekk1992) [reply src/main/java/com/example/moviemanager/configuration/TmdbConfigProperties.java]: Changed to record.

File: src/main/java/com/example/moviemanager/service/TmdbMovieService.java (hunk 22 above)
Comment (marekk1992) [reply src/main/java/com/example/moviemanager/service/TmdbMovieService.java]: Removed.

File: src/main/java/com/example/moviemanager/service/TmdbClient.java [hunk 23]
```diff
@@ -0,0 +1,58 @@
...
+
+    private final TmdbConfigProperties tmdbConfigProperties;
+    private final RestTemplate restTemplate;
+
+    public TmdbClient(TmdbConfigProperties tmdbConfigProperties, RestTemplate restTemplate) {
+        this.tmdbConfigProperties = tmdbConfigProperties;
+        this.restTemplate = restTemplate;
+    }
+
+    public MovieDetailsResponse find(FindMovieInfo findMovieInfo) {
```
Comment (RobertasVainickas) [inline src/main/java/com/example/moviemanager/service/TmdbClient.java]: When you call this method: `tmdbClient.find();` It is not really clear what you are trying to find, `findMovies` would be self explaining name

File: src/main/java/com/example/moviemanager/service/TmdbClient.java [hunk 24]
```diff
@@ -0,0 +1,58 @@
...
+                .newInstance()
+                .path(PATH + pathComponent)
+                .queryParam(API_KEY, tmdbConfigProperties.apiKey())
+                .queryParam(QUERY, findMovieInfo.title())
+                .queryParam(releaseYear, findMovieInfo.releaseYear())
+                .build()
+                .toUriString();
+    }
+
+    public MovieDetailsResponse getResponse(String endpoint) {
```
Comment (RobertasVainickas) [inline src/main/java/com/example/moviemanager/service/TmdbClient.java]: Why public?

File: src/test/java/com/example/moviemanager/service/TmdbClientTest.java [hunk 25]
```diff
@@ -0,0 +1,155 @@
...
+                        .withHeader(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE)
+                        .withStatus(HttpStatus.OK.value())
+                        .withBody(objectMapper.writeValueAsString(
+                                new MovieDetailsResponse(List.of(new MovieDetails("GOT Tv-show.", 8.4)))
+                        ))
+                )
+        );
+
+        // when
+        String actualMovieDetailsResponse = objectMapper.writeValueAsString(tmdbClient.find(findMovieInfo));
```
Comment (RobertasVainickas) [inline src/test/java/com/example/moviemanager/service/TmdbClientTest.java]: You are testing TmdbClient class not the 3rd party API. So why are you mapping your class response to json? It is not what you would expect to get as a response when you cal `find` method.

File: src/test/java/com/example/moviemanager/service/TmdbClientTest.java [hunk 26]
```diff
@@ -0,0 +1,155 @@
...
+        System.out.println(actualMovieDetailsResponse);
```
Comment (RobertasVainickas) [inline src/test/java/com/example/moviemanager/service/TmdbClientTest.java]: I think you forgot to delete it

File: src/test/java/com/example/moviemanager/service/TmdbClientTest.java [hunk 27]
```diff
@@ -0,0 +1,155 @@
...
+    }
+
+    @ParameterizedTest
+    @CsvFileSource(resources = "/invalid_find_movie_parameters.csv", numLinesToSkip = 1)
+    public void returns_empty_response_when_provided_non_existing_movie_parameters(
+            String title, MovieType type, int releaseYear
+    ) throws JsonProcessingException, JSONException {
+        // given
+        FindMovieInfo findMovieInfo = new FindMovieInfo(title, type, releaseYear);
+        String url = type.equals(MovieType.MOVIE) ? MOVIES_URL : TV_SHOWS_URL;
```
Comment (RobertasVainickas) [inline src/test/java/com/example/moviemanager/service/TmdbClientTest.java]: Instead of this, I suggest replacing MovieType in csv with actual URL. Since it is given part, you should not have any logic in it. It should be plain facts.

File: src/test/java/com/example/moviemanager/service/TmdbClientTest.java [hunk 28]
```diff
@@ -0,0 +1,155 @@
...
+
+        // then
+        JSONAssert.assertEquals(
+                """
+                          {   "results": [
+                              ]
+                          }
+                        """, actualMovieDetailsResponse, true);
+    }
+
```
Comment (RobertasVainickas) [inline src/test/java/com/example/moviemanager/service/TmdbClientTest.java]: You are missing test when response is null

File: src/main/java/com/example/moviemanager/service/MovieService.java [hunk 29]
```diff
@@ -49,4 +64,11 @@ public Movie update(UUID id, Movie movie) {
 
         return movieRepository.save(movie);
     }
-}
\ No newline at end of file
+
+    private MovieDetails retrieveMovieDetails(MovieDetailsResponse movieDetailsResponse) {
```
Comment (RobertasVainickas) [inline src/main/java/com/example/moviemanager/service/MovieService.java:68]: No test for it

**Review by RobertasVainickas — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/moviemanager/service/TmdbClient.java (hunk 23 above)
Comment (marekk1992) [reply src/main/java/com/example/moviemanager/service/TmdbClient.java]: Fixed.

File: src/main/java/com/example/moviemanager/service/TmdbClient.java (hunk 24 above)
Comment (marekk1992) [reply src/main/java/com/example/moviemanager/service/TmdbClient.java]: Fixed.

File: src/test/java/com/example/moviemanager/service/TmdbClientTest.java (hunk 26 above)
Comment (marekk1992) [reply src/test/java/com/example/moviemanager/service/TmdbClientTest.java]: Deleted.

File: src/test/java/com/example/moviemanager/service/TmdbClientTest.java (hunk 25 above)
Comment (marekk1992) [reply src/test/java/com/example/moviemanager/service/TmdbClientTest.java]: Fixed.

File: src/test/java/com/example/moviemanager/service/TmdbClientTest.java (hunk 27 above)
Comment (marekk1992) [reply src/test/java/com/example/moviemanager/service/TmdbClientTest.java]: Included url in .csv file, but still kept MovieType, since it is required to buid FindMovieInfo object.

File: src/test/java/com/example/moviemanager/service/TmdbClientTest.java (hunk 28 above)
Comment (marekk1992) [reply src/test/java/com/example/moviemanager/service/TmdbClientTest.java]: Not required. Since in normal flow response can`t be null (either exception by RestTemplate is thrown, or empty response is returned).

File: src/main/java/com/example/moviemanager/service/MovieService.java (hunk 29 above)
Comment (marekk1992) [reply src/main/java/com/example/moviemanager/service/MovieService.java:68]: created additional test, for covering cases when MovieDetailsResponse is empty and when contains more than one movie.

File: src/test/java/com/example/moviemanager/service/TmdbClientTest.java [hunk 30]
```diff
@@ -0,0 +1,121 @@
...
+import java.util.Collections;
+import java.util.List;
+
+import static com.github.tomakehurst.wiremock.client.WireMock.aResponse;
+import static com.github.tomakehurst.wiremock.client.WireMock.get;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TmdbClientTest {
+
+    private static final String TV_SHOWS_URL = "/3/search/tv?api_key=43d5d872d7dc32845ab997b8v25987f7" +
```
Comment (RobertasVainickas) [inline src/test/java/com/example/moviemanager/service/TmdbClientTest.java]: I don't like that both urls contain query information specific for one specific test. Either have only base url together with api key here (then you at least have it reusable) or have the whole url in test.

**Review by RobertasVainickas — APPROVED** [review]
